File: 20th/코딩테스트 심화반/week1/2580번_스도쿠.cpp
#include <iostream>
#include <vector>

using namespace std;

int board[9][9];
vector<pair<int, int>> v; // 빈 칸의 좌표(x, y) 저장

// 백트래킹
bool back_tracking(int idx) {
	if (v.size() == idx) return true; // 마지막까지 빈 칸을 채웠을 경우 true 반환

	int i = v[idx].first;
	int j = v[idx].second;
	int area_i = i / 3 * 3; // 9개의 3 by 3 정사각형 중 행의 위치
	int area_j = j / 3 * 3; // 9개의 3 by 3 정사각형 중 열의 위치
	for (int num = 1; num <= 9; num++) {
		bool possible = true;
		// 가로 or 세로에 num이 존재할 경우 불가능
		for (int k = 1; k <= 9; k++) {
			if (board[(i + k) % 9][j] == num) possible = false;
			if (board[i][(j + k) % 9] == num) possible = false;
		}
		// 3 by 3 정사각형 내에 num이 존재할 경우 불가능
		for (int r = area_i; r < area_i + 3; r++) {
			for (int c = area_j; c < area_j + 3; c++) {
				if (board[r][c] == num) possible = false;
			}
		}

		if (possible) {
			board[i][j] = num; // 보드에 num을 넣음
			if (back_tracking(idx + 1)) return true;
			board[i][j] = 0; // 해당 num을 넣어서 실패했을 경우 다시 0으로 초기화
		}
	}
	
	return false;
}

int main() {
	for (int i = 0; i < 9; i++) {
		for (int j = 0; j < 9; j++) {
			cin >> board[i][j];
			if (board[i][j] == 0) {
				v.push_back({ i, j });
			}
		}
	}

	// 스도쿠를 해결했을 경우
	if (back_tracking(0)) {
		for (int i = 0; i < 9; i++) {
			for (int j = 0; j < 9; j++) {
				cout << board[i][j] << " ";
			}
			cout << "\n";
		}
	}
}

File: 20th/코딩테스트 심화반/week1/2580번_스도쿠.py
v = []

def back_tracking(idx):
    global sudoku
    
    # 모든 빈칸을 채웠으면 종료
    if idx == len(v):
        return True
    
    i, j = v[idx]
    area_i = i // 3 * 3
    area_j = j // 3 * 3

    for n in range(1, 10):
        possible = True
        
        # 행, 열 검사
        for m in range(9):
            if sudoku[i][m] == n or sudoku[m][j] == n:
                possible = False
                break
        
        # 3x3 구역 검사
        for r in range(area_i, area_i + 3):
            for c in range(area_j, area_j + 3):
                if sudoku[r][c] == n:
                    possible = False
                    break
            if not possible:
                break
        
        if possible:
            sudoku[i][j] = n
            if back_tracking(idx + 1):
                return True
            sudoku[i][j] = 0  # 백트래킹 되돌리기
    
    return False

# 입력 받기
sudoku = [list(map(int, input().split())) for _ in range(9)]

# 빈칸 좌표 저장
for i in range(9):
    for j in range(9):
        if sudoku[i][j] == 0:
            v.append((i, j))

# 백트래킹 수행
back_tracking(0)

# 결과 출력
for row in sudoku:
    print(" ".join(map(str, row)))
